File: hdf-diff/src/renderers/markdown.ts
import type { HdfComparison, RequirementDiff } from '../types.js';
import type { RenderOptions } from './types.js';
import { filterRequirements } from './filter.js';

/**
 * Render the summary table in markdown format.
 */
function renderSummaryTable(comparison: HdfComparison): string {
  const { summary } = comparison;
  const lines: string[] = [
    '## HDF Comparison Summary',
    '',
    '| Metric | Count |',
    '|--------|-------|',
    `| Fixed | ${summary.fixed} |`,
    `| Regressed | ${summary.regressed} |`,
    `| New | ${summary.new} |`,
    `| Absent | ${summary.absent} |`,
    `| Unchanged | ${summary.unchanged} |`,
    `| Updated | ${summary.updated} |`,
    `| **Total** | **${summary.total}** |`,
  ];
  return lines.join('\n');
}

/**
 * Group requirement diffs by state.
 */
function groupByState(
  diffs: RequirementDiff[],
): Map<string, RequirementDiff[]> {
  const groups = new Map<string, RequirementDiff[]>();
  for (const diff of diffs) {
    const existing = groups.get(diff.state);
    if (existing) {
      existing.push(diff);
    } else {
      groups.set(diff.state, [diff]);
    }
  }
  return groups;
}

/**
 * Escape a value for use in a markdown table cell.
 * Replaces pipe characters with their HTML entity.
 */
function escapeCell(value: string): string {
  return value.replace(/\|/g, '&#124;');
}

/**
 * Format field changes for a markdown cell.
 */
function formatFieldChanges(req: RequirementDiff): string {
  if (req.fieldChanges.length === 0) return '';
  return req.fieldChanges
    .map((fc) => {
      if (fc.op === 'add') return `+${fc.path}: ${JSON.stringify(fc.newValue)}`;
      if (fc.op === 'remove')
        return `-${fc.path}: ${JSON.stringify(fc.oldValue)}`;
      return `${fc.path}: ${JSON.stringify(fc.oldValue)} -> ${JSON.stringify(fc.newValue)}`;
    })
    .join('; ');
}

/**
 * Render a section for a single state group.
 */
function renderStateSection(
  state: string,
  diffs: RequirementDiff[],
  detail: string,
): string {
  const label = state.charAt(0).toUpperCase() + state.slice(1);
  const lines: string[] = [`### ${label} (${diffs.length})`];

  if (diffs.length === 0) {
    lines.push('', '(none)');
    return lines.join('\n');
  }

  lines.push('');

  if (detail === 'full') {
    lines.push(
      '| ID | Title | Old Status | New Status | Change Reasons | Field Changes |',
      '|----|-------|------------|------------|----------------|---------------|',
    );
    for (const req of diffs) {
      const id = escapeCell(req.id);
      const title = escapeCell(req.title ?? '');
      const oldStatus = escapeCell(req.oldEffectiveStatus ?? '');
      const newStatus = escapeCell(req.newEffectiveStatus ?? '');
      const reasons = escapeCell(req.changeReasons.join(', '));
      const fieldChanges = escapeCell(formatFieldChanges(req));
      lines.push(
        `| ${id} | ${title} | ${oldStatus} | ${newStatus} | ${reasons} | ${fieldChanges} |`,
      );
    }
  } else {
    lines.push(
      '| ID | Title | Old Status | New Status |',
      '|----|-------|------------|------------|',
    );
    for (const req of diffs) {
      const id = escapeCell(req.id);
      const title = escapeCell(req.title ?? '');
      const oldStatus = escapeCell(req.oldEffectiveStatus ?? '');
      const newStatus = escapeCell(req.newEffectiveStatus ?? '');
      lines.push(`| ${id} | ${title} | ${oldStatus} | ${newStatus} |`);
    }
  }

  return lines.join('\n');
}

/**
 * Render an HdfComparison as a Markdown string.
 *
 * - `detail: 'summary'` -- summary table only
 * - `detail: 'control'` -- summary + per-requirement tables by state
 * - `detail: 'full'`    -- summary + per-requirement tables with changeReasons and fieldChanges
 *
 * Default detail level: `'control'`.
 */
export function renderMarkdown(
  comparison: HdfComparison,
  options?: RenderOptions,
): string {
  const detail = options?.detail ?? 'control';
  const parts: string[] = [renderSummaryTable(comparison)];

  if (detail === 'summary') {
    return parts.join('\n');
  }

  // Determine which states to display
  const stateOrder: string[] = [
    'fixed',
    'regressed',
    'new',
    'absent',
    'updated',
    'unchanged',
  ];

  const filtered = filterRequirements(
    comparison.requirementDiffs,
    options,
  );
  const grouped = groupByState(filtered);

  // If filtering by states, only show those specific states
  const statesToShow =
    options?.filterStates && options.filterStates.length > 0
      ? stateOrder.filter((s) => options.filterStates!.includes(s))
      : stateOrder;

  for (const state of statesToShow) {
    parts.push('');
    const diffs = grouped.get(state) ?? [];
    parts.push(renderStateSection(state, diffs, detail));
  }

  return parts.join('\n');
}

File: hdf-cli/pkg/diff/render/markdown.go
package render

import (
	"fmt"
	"strings"

	types "github.com/mitre/hdf-cli/pkg/diff/types"
)

// stateOrder defines the display order for requirement state sections.
var stateOrder = []types.RequirementState{
	types.StateFixed,
	types.StateRegressed,
	types.StateNew,
	types.StateAbsent,
	types.StateUpdated,
	types.StateUnchanged,
}

// Markdown renders an HdfComparison as a Markdown string.
//
//   - DetailSummary: summary table only
//   - DetailControl: summary + per-requirement tables by state
//   - DetailFull: summary + per-requirement tables with changeReasons and fieldChanges
func Markdown(comparison types.HdfComparison, opts Options) (string, error) {
	detail := opts.effectiveDetail()

	var sb strings.Builder

	renderSummaryTable(&sb, comparison)

	if detail == DetailSummary {
		return sb.String(), nil
	}

	filtered := filterRequirements(comparison.RequirementDiffs, opts)
	grouped := groupByState(filtered)

	// Determine which states to show
	statesToShow := stateOrder
	if len(opts.FilterStates) > 0 {
		allowed := make(map[types.RequirementState]bool, len(opts.FilterStates))
		for _, s := range opts.FilterStates {
			allowed[s] = true
		}
		var selected []types.RequirementState
		for _, s := range stateOrder {
			if allowed[s] {
				selected = append(selected, s)
			}
		}
		statesToShow = selected
	}

	for _, state := range statesToShow {
		sb.WriteString("\n")
		diffs := grouped[state]
		renderStateSection(&sb, state, diffs, detail)
	}

	return sb.String(), nil
}

// renderSummaryTable writes the markdown summary table.
func renderSummaryTable(sb *strings.Builder, comparison types.HdfComparison) {
	s := comparison.Summary
	sb.WriteString("## HDF Comparison Summary\n")
	sb.WriteString("\n")
	sb.WriteString("| Metric | Count |\n")
	sb.WriteString("|--------|-------|\n")
	fmt.Fprintf(sb, "| Fixed | %d |\n", s.Fixed)
	fmt.Fprintf(sb, "| Regressed | %d |\n", s.Regressed)
	fmt.Fprintf(sb, "| New | %d |\n", s.New)
	fmt.Fprintf(sb, "| Absent | %d |\n", s.Absent)
	fmt.Fprintf(sb, "| Unchanged | %d |\n", s.Unchanged)
	fmt.Fprintf(sb, "| Updated | %d |\n", s.Updated)
	fmt.Fprintf(sb, "| **Total** | **%d** |", s.Total)
}

// groupByState groups requirement diffs by their state.
func groupByState(diffs []types.RequirementDiff) map[types.RequirementState][]types.RequirementDiff {
	groups := make(map[types.RequirementState][]types.RequirementDiff)
	for _, d := range diffs {
		groups[d.State] = append(groups[d.State], d)
	}
	return groups
}

// renderStateSection writes a markdown section for a single state group.
func renderStateSection(sb *strings.Builder, state types.RequirementState, diffs []types.RequirementDiff, detail DetailLevel) {
	label := strings.ToUpper(string(state)[:1]) + string(state)[1:]
	fmt.Fprintf(sb, "### %s (%d)\n", label, len(diffs))

	if len(diffs) == 0 {
		sb.WriteString("\n(none)")
		return
	}

	sb.WriteString("\n")

	if detail == DetailFull {
		sb.WriteString("| ID | Title | Old Status | New Status | Change Reasons | Field Changes |\n")
		sb.WriteString("|----|-------|------------|------------|----------------|---------------|\n")
		for _, req := range diffs {
			id := escapeMarkdownCell(req.ID)
			title := escapeMarkdownCell(req.Title)
			oldStatus := escapeMarkdownCell(req.OldEffectiveStatus)
			newStatus := escapeMarkdownCell(req.NewEffectiveStatus)
			reasons := escapeMarkdownCell(formatChangeReasons(req.ChangeReasons))
			fieldChanges := escapeMarkdownCell(formatFieldChangesWithArrow(req.FieldChanges, "->"))
			fmt.Fprintf(sb, "| %s | %s | %s | %s | %s | %s |\n", id, title, oldStatus, newStatus, reasons, fieldChanges)
		}
	} else {
		sb.WriteString("| ID | Title | Old Status | New Status |\n")
		sb.WriteString("|----|-------|------------|------------|\n")
		for _, req := range diffs {
			id := escapeMarkdownCell(req.ID)
			title := escapeMarkdownCell(req.Title)
			oldStatus := escapeMarkdownCell(req.OldEffectiveStatus)
			newStatus := escapeMarkdownCell(req.NewEffectiveStatus)
			fmt.Fprintf(sb, "| %s | %s | %s | %s |\n", id, title, oldStatus, newStatus)
		}
	}
}

// escapeMarkdownCell replaces pipe characters with HTML entities.
func escapeMarkdownCell(value string) string {
	return strings.ReplaceAll(value, "|", "&#124;")
}
